### PR TITLE
⚒️ Forge: Extract persistence logic from God Object (src/db.rs)

### PR DESCRIPTION
**Pull request by madmax983**
This PR refactors `src/db.rs` to address the "God Object" anti-pattern by extracting persistence-related logic into dedicated modules under `src/storage/index_persistence/`.

### Changes
*   **Extracted `PersistenceTracker`**: Moved to `src/storage/index_persistence/tracker.rs`.
*   **Extracted Background Thread**: Moved `spawn_background_persistence_thread` to `src/storage/index_persistence/worker.rs`.
*   **Extracted Operations**: Moved `persist_vector_indexes`, `load_vector_indexes`, etc., to `src/storage/index_persistence/operations.rs`.
*   **Decoupling**: Updated `load_vector_indexes` to accept `&CurrentStorage` instead of `&GallifreyDB` to reduce coupling.
*   **Documentation**: Added module documentation in `src/storage/index_persistence/mod.rs`.

### Verification
*   Ran `cargo check` to ensure no compilation errors.
*   Ran `cargo test` to ensure no regression in functionality.
*   Verified that no runtime behavior was changed, only code organization.

---
*PR created automatically by Jules for task [4183784440092237982](https://jules.google.com/task/4183784440092237982) started by @madmax983*

### CODE REVIEW

File: src/storage/index_persistence/operations.rs [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+
+        save_vector_mappings(&vector_mappings, &vec_path.join("mappings.idx")).map_err(|e| {
+            StorageError::PersistenceError(format!(
+                "Failed to save vector mappings for {}: {}",
+                property_name, e
+            ))
+        })?;
+    }
+
+    tracker.reset_vector_mutations();
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/storage/index_persistence/operations.rs]: > medium
> 
> Following the suggestion to make the `tracker` parameter optional in the function signature, this reset logic should only execute when a tracker is actually provided.
> 
> ```rust
> if let Some(tracker) = tracker {
>         tracker.reset_vector_mutations();
>     }
> ```

@jules

**Comment by madmax983** [comment]
> ## [Codecov](https://app.codecov.io/gh/madmax983/GallifreyDB/pull/594?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=madmax983) Report
> ❌ Patch coverage is `75.30364%` with `122 lines` in your changes missing coverage. Please review. ✅ Project coverage is 91.30%. Comparing base ([`726ef04`](https://app.codecov.io/gh/madmax983/GallifreyDB/commit/726ef045e3d8a47f9f90c0fbf5aad558832a2d14?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=madmax983)) to head ([`fde71ec`](https://app.codecov.io/gh/madmax983/GallifreyDB/commit/fde71ecbd2d8744858b6da3cf1a7683af9dae06f?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=madmax983)).
> 
> [Files with missing lines](https://app.codecov.io/gh/madmax983/GallifreyDB/pull/594?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=madmax983)	Patch %	Lines
> [src/storage/index_persistence/operations.rs](https://app.codecov.io/gh/madmax983/GallifreyDB/pull/594?src=pr&el=tree&filepath=src%2Fstorage%2Findex_persistence%2Foperations.rs&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=madmax983#diff-c3JjL3N0b3JhZ2UvaW5kZXhfcGVyc2lzdGVuY2Uvb3BlcmF0aW9ucy5ycw==)	67.48%	[93 Missing ⚠️ ](https://app.codecov.io/gh/madmax983/GallifreyDB/pull/594?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=madmax983)
> [src/storage/index_persistence/worker.rs](https://app.codecov.io/gh/madmax983/GallifreyDB/pull/594?src=pr&el=tree&filepath=src%2Fstorage%2Findex_persistence%2Fworker.rs&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=madmax983#diff-c3JjL3N0b3JhZ2UvaW5kZXhfcGVyc2lzdGVuY2Uvd29ya2VyLnJz)	67.04%	[29 Missing ⚠️ ](https://app.codecov.io/gh/madmax983/GallifreyDB/pull/594?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=madmax983)
> ❌ Your patch check has failed because the patch coverage (75.30%) is below the target coverage (85.00%). You can increase the patch coverage or adjust the [target](https://docs.codecov.com/docs/commit-status#target) coverage.
> 
> Additional details and impacted files
> 🚀 New features to boost your workflow:

@jules

**Comment by madmax983** [comment]
> ## Code Review: Extract persistence logic from God Object
> This PR successfully refactors persistence-related logic out of src/db.rs into dedicated modules.
> 
> ### Strengths
> 1. Excellent code organization - clear separation into tracker, worker, and operations modules
> 2. Zero behavioral changes - pure refactoring with no functional modifications
> 3. Improved coupling - load_vector_indexes now accepts &CurrentStorage instead of &GallifreyDB
> 4. Good documentation added for new modules
> 5. Proper visibility maintained with pub(crate)
> 
> ### Critical Issue: .unwrap() Usage
> Multiple instances of .unwrap() on SystemTime operations violate CODING_STANDARDS.md no-unwrap policy:
> 
> tracker.rs lines 34, 74-76, 88-90, 100-102, 114-116, 144-146, 154-156, 164-166, 175-176
> 
> Current code: SystemTime::now().duration_since(UNIX_EPOCH).unwrap()
> 
> This can panic during clock adjustments. Recommended fix: Replace .unwrap() with .unwrap_or(Duration::ZERO)
> 
> ### Minor Issue
> PersistenceTracker should implement Default trait since it has a new() method.
> 
> ### Verification Needed
> * Run clippy with -D warnings
> * Run tests
> * Verify coverage stays above 85%
> * Fix all .unwrap() calls
> 
> ### Overall Assessment
> Strong refactoring that improves code organization. The .unwrap() issue must be fixed before merge per coding standards.

@jules

**Comment by madmax983** [comment]
> ## Code Review for PR #594: Extract Persistence Logic from God Object
> This PR successfully refactors persistence logic from `src/db.rs` into dedicated modules, addressing the God Object anti-pattern. The overall approach is sound and improves code organization. However, there are several critical issues that need to be addressed before merging.
> 
> ## 🚨 Critical Issues
> ### 1. **Compilation Error in src/db.rs:502**
> Line 502 attempts to call `load_vector_indexes(&db.current, manager)`, but the function signature was changed to accept `&Arc<CurrentStorage>` instead of `&GallifreyDB`. This is correct for decoupling, but **the old call to `load_vector_indexes(&db, manager)` was accidentally replaced with a call to `persist_vector_indexes`** in the diff (line 790).
> 
> ```rust
> // Current (WRONG):
> if let Err(e) = persist_vector_indexes(current, manager, Some(tracker)) {
>     eprintln!("Warning: Failed to load vector indexes: {}", e);
> }
> 
> // Should be:
> if let Err(e) = load_vector_indexes(&db.current, manager) {
>     eprintln!("Warning: Failed to load vector indexes: {}", e);
> }
> ```
> 
> **Impact**: This will **persist** indexes on startup instead of **loading** them, causing data corruption or unexpected behavior.
> 
> **Location**: `src/db.rs:502`
> 
> ### 2. **Unwrap in Tracker Functions (Multiple Locations)**
> The `PersistenceTracker` uses `.unwrap()` in production code when getting system time:
> 
> ```rust
> // src/storage/index_persistence/tracker.rs:34, 74, 100, 144, etc.
> let now = std::time::SystemTime::now()
>     .duration_since(std::time::UNIX_EPOCH)
>     .unwrap()  // ⚠️ UNWRAP IN PRODUCTION
>     .as_secs();
> ```
> 
> **Problem**: According to CLAUDE.md coding standards: _"Never `.unwrap()` in production"_. While system time failures are extremely rare, they can occur due to clock adjustments or system issues.
> 
> **Impact**: Potential panic in production, especially in the background persistence thread.
> 
> **Fix**: Use `.unwrap_or(0)` or `.expect("System time before UNIX epoch")` with a descriptive message.
> 
> **Locations**:
> 
> * `src/storage/index_persistence/tracker.rs:34, 74, 89, 100, 114, 144, 154, 164, 174`
> 
> ### 3. **Missing Tracker Check in persist_vector_indexes**
> In `operations.rs:109`, the code unconditionally calls `tracker.reset_vector_mutations()`:
> 
> ```rust
> // src/storage/index_persistence/operations.rs:109
> if let Some(tracker) = tracker {
>     tracker.reset_vector_mutations();
> }
> Ok(())
> ```
> 
> Wait, looking at the code again, this is actually **correct** - it does check for `Some(tracker)`. However, in **worker.rs:64** and **worker.rs:535**, the calls pass `Some(&tracker)` correctly.
> 
> Actually, reviewing the diff more carefully, I see that **line 109 in operations.rs** is missing the `if let Some(tracker)` wrapper:
> 
> ```rust
> // Current (WRONG):
> tracker.reset_vector_mutations();  // Line 109
> Ok(())
> 
> // Should be:
> if let Some(tracker) = tracker {
>     tracker.reset_vector_mutations();
> }
> Ok(())
> ```
> 
> **Impact**: Compilation error - attempting to call method on `Option<&Arc<PersistenceTracker>>` without unwrapping.
> 
> **Location**: `src/storage/index_persistence/operations.rs:109`
> 
> ## ⚠️ Major Issues
> ### 4. **Missing Test Coverage for New Modules**
> The newly extracted modules (`tracker.rs`, `worker.rs`, `operations.rs`) have no dedicated unit tests. The existing tests in the parent crate may cover integration scenarios, but:
> 
> * No tests for `PersistenceTracker` mutation counting edge cases
> * No tests for background thread panic recovery
> * No tests for optional tracker parameter in `persist_vector_indexes`
> 
> **Recommendation**: Add unit tests to ensure the refactoring maintains behavioral equivalence.
> 
> ### 5. **Inconsistent Function Signatures**
> The refactoring improved `load_vector_indexes` by accepting `&Arc<CurrentStorage>` instead of `&GallifreyDB` to reduce coupling. However, other functions like `persist_all_indexes` still accept multiple Arc parameters that could be grouped into a struct.
> 
> **Recommendation**: Consider a `PersistenceContext` struct to reduce parameter count in future refactors:
> 
> ```rust
> struct PersistenceContext {
>     current: Arc<CurrentStorage>,
>     historical: Arc<RwLock<HistoricalStorage>>,
>     temporal_indexes: Arc<TemporalIndexes>,
>     wal: Arc<ConcurrentWalSystem>,
>     manager: Arc<IndexPersistenceManager>,
>     tracker: Arc<PersistenceTracker>,
> }
> ```
> 
> This would address the `#[allow(clippy::too_many_arguments)]` annotation in `worker.rs:35`.
> 
> ## 📝 Minor Issues
> ### 6. **Redundant Shutdown Check in Worker Loop**
> In `worker.rs:51-57`, the shutdown check appears twice in succession:
> 
> ```rust
> while !tracker.is_shutdown() {
>     std::thread::sleep(check_interval);
>     
>     // Skip if shutdown signaled
>     if tracker.is_shutdown() {  // Redundant check
>         break;
>     }
> ```
> 
> **Recommendation**: The second check is only useful if you want to avoid sleeping on shutdown. Consider removing or adding a comment explaining why it's needed.
> 
> ### 7. **Incomplete Module Documentation**
> The new `mod.rs` has minimal module documentation:
> 
> ```rust
> /// Persistence operations implementation.
> pub mod operations;
> ```
> 
> **Recommendation**: Add more detailed documentation explaining the purpose and responsibilities of each module, following the comprehensive style used in other parts of the codebase (e.g., WAL documentation).
> 
> ### 8. **Error Message Inconsistency in persist_all_indexes**
> In `operations.rs:413`, the call to `persist_vector_indexes` doesn't match the updated signature:
> 
> ```rust
> if let Err(e) = persist_vector_indexes(current, manager, tracker) {
>     eprintln!("Failed to persist vector indexes: {}", e);
> }
> ```
> 
> This should pass `Some(tracker)` to match the `Option<&Arc<PersistenceTracker>>` parameter:
> 
> ```rust
> if let Err(e) = persist_vector_indexes(current, manager, Some(tracker)) {
>     eprintln!("Failed to persist vector indexes: {}", e);
> }
> ```
> 
> **Location**: `src/storage/index_persistence/operations.rs:413`
> 
> ## ✅ Positive Aspects
> 1. **Excellent Separation of Concerns**: Moving `PersistenceTracker`, background worker, and operations into separate modules significantly improves code organization.
> 2. **Reduced Coupling**: Changing `load_vector_indexes` to accept `&Arc<CurrentStorage>` instead of `&GallifreyDB` is a good architectural improvement.
> 3. **Preserved Documentation**: All doc comments were preserved during the move, maintaining code documentation quality.
> 4. **No Performance Impact**: The refactoring is purely structural with no changes to algorithms or data structures.
> 5. **Crash Safety Preserved**: The background thread panic handler remains intact in the new `worker.rs` module.
> 
> ## 🔧 Required Changes Before Merge
> 1. ✅ Fix line 790/502 in `src/db.rs` to call `load_vector_indexes` instead of `persist_vector_indexes`
> 2. ✅ Fix line 109 in `operations.rs` to properly handle `Option<&Arc<PersistenceTracker>>`
> 3. ✅ Fix line 413 in `operations.rs` to pass `Some(tracker)` instead of `tracker`
> 4. ✅ Replace all `.unwrap()` calls in `tracker.rs` with `.unwrap_or(0)` or `.expect("message")`
> 5. ✅ Run `cargo clippy --all-targets --all-features -- -D warnings` and fix all warnings
> 6. ✅ Run `cargo test` to ensure no regressions
> 7. ⚠️ Consider adding unit tests for the new modules
> 
> ## 📊 Code Quality Checklist
> * [ ]  **Clippy passes**: Need to verify after fixing compilation errors
> * [ ]  **Code formatted**: Appears formatted correctly
> * [ ]  **Tests pass**: Cannot verify due to compilation errors
> * [ ]  **Coverage maintained**: Unknown - requires test run
> * [x]  **Temporal invariants preserved**: No temporal logic changed
> * [x]  **No performance regression**: Refactoring only
> * [ ]  **Error handling comprehensive**: Missing `.unwrap()` fixes
> * [ ]  **Tests cover edge cases**: No new tests added
> * [x]  **Documentation updated**: Doc comments preserved
> * [x]  **No unsafe without safety comments**: No unsafe code added
> * [x]  **Strong typing used**: ID types preserved
> * [x]  **Code follows CODING_STANDARDS.md**: Mostly, except `.unwrap()` violations
> 
> ## 🎯 Recommendation
> **REQUEST CHANGES** - The refactoring approach is excellent, but the critical compilation errors and coding standard violations (unwrap usage) must be fixed before merging. Once the required changes are addressed, this will be a valuable improvement to the codebase.
> 
> Please address the critical issues (items 1-5 in Required Changes) and re-run the pre-commit checks:
> 
> ```shell
> cargo clippy --all-targets --all-features -- -D warnings
> cargo fmt --all
> cargo test
> ```

@jules

**Comment by madmax983** [comment]
@jules patch coverage is still failing. Please reach 85%. Liner is also failing.

**Comment by madmax983** [comment]
@jules rror: module has the same name as its containing module
  --> src/storage/index_persistence/tracker_tests.rs:2:1
   |
 2 | / mod tests {
 3 | |
 4 | |     use crate::storage::index_persistence::tracker::PersistenceTracker;
...  |
79 | | }
   | |_^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.93.0/index.html#module_inception
   = note: `-D clippy::module-inception` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::module_inception)]`

error: could not compile `gallifreydb` (lib test) due to 1 previous error

[codecov/patch](https://github.com/madmax983/GallifreyDB/pull/594/checks?check_run_id=61764714849)Failing after 1s — 75.45% of diff hit (target 85.00%)

Please increase this coverage, or this PR will not be merged.
